vllm/worker/habana_model_runner.py

```diff
@@ -1832,6 +1789,60 @@ def _check_config(self, batch_size, seq_len, is_prompt, warmup_mode):
             logger.warning("Configuration: (%s, %s, %s) was not warmed-up!",
                            phase, batch_size, seq_len)
 
+    def create_lora_mask(self, input_tokens: torch.Tensor, lora_ids: List[int], is_prompt: bool):
+        lora_mask: torch.Tensor = None
+        lora_logits_mask: torch.Tensor = None
+        lora_index = 0
+
+        if is_prompt:
+            lora_mask = torch.zeros(
+                input_tokens.shape[0] * input_tokens.shape[1],
+                (self.lora_config.max_loras) * self.lora_config.max_lora_rank,
+                dtype=self.lora_config.lora_dtype)
+            lora_logits_mask = torch.zeros(input_tokens.shape[0],
+                                           (self.lora_config.max_loras) *
+                                           self.lora_config.max_lora_rank,
+                                           dtype=self.lora_config.lora_dtype)
+
+            ones = torch.ones(input_tokens.shape[1],
+                              self.lora_config.max_lora_rank,
+                              dtype=self.lora_config.lora_dtype)
+            logit_ones = torch.ones(1,
+                                    self.lora_config.max_lora_rank,
+                                    dtype=self.lora_config.lora_dtype)
+
+            for i in range(len(lora_ids)):
+                if lora_ids[i] == 0:
+                    continue
+                lora_index = self.lora_manager._adapter_manager.lora_index_to_id.index(lora_ids[i])
+                start_row = i * input_tokens.shape[1]
+                end_row = start_row + input_tokens.shape[1]
+                start_col = lora_index * self.lora_config.max_lora_rank
+                end_col = start_col + self.lora_config.max_lora_rank
+                lora_mask[start_row:end_row, start_col:end_col] = ones
+                lora_logits_mask[i, start_col:end_col] = logit_ones
+            lora_mask = lora_mask.to('hpu')
+            lora_logits_mask = lora_logits_mask.to('hpu')
+        else:
+            lora_mask = torch.zeros(input_tokens.shape[0],
+                                    (self.lora_config.max_loras) *
+                                    self.lora_config.max_lora_rank,
+                                    dtype=self.lora_config.lora_dtype)
+            ones = torch.ones(1,
+                              self.lora_config.max_lora_rank,
+                              dtype=self.lora_config.lora_dtype)
+            for i in range(len(lora_ids)):
+                if lora_ids[i] == 0:
+                    continue
+                lora_index = self.lora_manager._adapter_manager.lora_index_to_id.index(lora_ids[i])
+                start_pos = lora_index * self.lora_config.max_lora_rank
+                end_pos = start_pos + self.lora_config.max_lora_rank
+                lora_mask[i, start_pos:end_pos] = ones
+            lora_mask = lora_mask.to('hpu')
+            lora_logits_mask = lora_mask
```

Comment:
Could you explain why now logits mask is pointing to mask, before it was left None for decode phase.
Is that because now in line 1906 there can be reference to an object that is None? If so I would rather add there check if lora_logits_mask is not none instead of changing the logic of what lora_logits_mask is keeping

Reply:
@michalkuligowski for decode phase lora_mask and logits_mask are same; previously we were setting it at execute_model; just moved to create_lora_mask for a cleaner code